syntax = "proto3";

package chord;

service ChordService {
    rpc GetSuccessor (ID) returns (NodeInfo);
    rpc GetPredecessor (Empty) returns (NodeInfo);
    rpc UpdatePredecessor (NodeInfo) returns (Empty);
    rpc Ping (Empty) returns (Empty);


    rpc Get (Key) returns (Value);
    rpc Put (KeyValue) returns (Empty);
    rpc Delete (Key) returns (Empty);

    rpc GetAllKeys (Empty) returns (KeyValueList);

    rpc SetPartition (Partition) returns (Ack);
    rpc ResolveData (Partition) returns (PartitionResult); 
}


message Empty {}


message ID {
    uint64 id = 1;
}


message Key {
    string key = 1;
}


message Value {
    string value = 1;
}


message KeyValue {
    string key = 1;
    string value = 2;
}

message KeyValueList {
    repeated KeyValue items = 1;
}

message NodeInfo {
    uint64 id = 1;
    string address = 2;
}

message Ack {
    bool ok = 1;
}

message Partition {
    map<string, string> values = 1;
    map<string, uint64> versions = 2;
    map<string, uint64> removed = 3;
}

message PartitionResult {
    bool ok = 1;
    Partition partition = 2;
}